local-cluster/tests/local_cluster.rs

```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
+
+    validator_a_info.config.wait_for_supermajority = Some(hard_fork_slot);
+    validator_a_info.config.expected_shred_version = Some(expected_shred_version);
+
+    validator_b_info.config.wait_for_supermajority = Some(hard_fork_slot);
+    validator_b_info.config.expected_shred_version = Some(expected_shred_version);
+
+    restart_whole_cluster_after_hard_fork(
+        &cluster,
+        validator_a_pubkey,
+        validator_b_pubkey,
+        validator_a_info,
+        validator_b_info,
+    );
+    // new slots should be rooted after hard-fork cluster relaunch
+    cluster
+        .lock()
+        .unwrap()
+        .check_for_new_roots(16, "hard fork", SocketAddrSpace::Unspecified);
+
+    // drop everything to open blockstores below
+    drop(cluster);
+
+    let (common_last_vote, common_root) = {
+        let (last_vote_a, _) = last_vote_in_tower(&val_a_ledger_path, &validator_a_pubkey).unwrap();
+        let (last_vote_b, _) = last_vote_in_tower(&val_b_ledger_path, &validator_b_pubkey).unwrap();
+        let root_a = root_in_tower(&val_a_ledger_path, &validator_a_pubkey).unwrap();
+        let root_b = root_in_tower(&val_b_ledger_path, &validator_b_pubkey).unwrap();
+        (last_vote_a.min(last_vote_b), root_a.min(root_b))
+    };
+
+    let blockstore_a = Blockstore::open(&val_a_ledger_path).unwrap();
+    let blockstore_b = Blockstore::open(&val_b_ledger_path).unwrap();
+
+    // collect all slots
+    let slots_a = AncestorIterator::new(common_last_vote, &blockstore_a).collect::<Vec<_>>();
+    let roots_a = blockstore_a
+        .reversed_rooted_slot_iterator(common_root)
+        .unwrap()
+        .collect::<Vec<_>>();
+    let slots_b = AncestorIterator::new(common_last_vote, &blockstore_b).collect::<Vec<_>>();
+    let roots_b = blockstore_b
+        .reversed_rooted_slot_iterator(common_root)
+        .unwrap()
+        .collect::<Vec<_>>();
+
+    // compare all slots
+    if !with_gap {
+        assert_eq!((slots_a, roots_a), (slots_b, roots_b));
+    } else {
+        // rough way to detect a gap with mostly similar slots...
+        assert_eq!(
+            (slots_a.first().unwrap(), roots_a.first().unwrap()),
+            (slots_b.first().unwrap(), roots_b.first().unwrap())
+        );
+        assert_eq!(
+            (slots_a.last().unwrap(), roots_a.last().unwrap()),
+            (slots_b.last().unwrap(), roots_b.last().unwrap())
+        );
+        assert_ne!((slots_a, roots_a), (slots_b, roots_b));
+    }
+}
+
+// the following two tests are rather fragile, depending on successful loading of the hard-forked snapshot.
+// so, make sure they aren't broken by testing (expected) failures as well with those conditioned
+// two actual test code invocations.
+#[test]
+#[serial]
+fn test_hard_fork_without_gap_in_roots() {
+    do_test_hard_fork_with_or_without_gap_in_roots(false);
+}
+
+#[test]
+#[serial]
+fn test_hard_fork_with_gap_in_roots() {
+    do_test_hard_fork_with_or_without_gap_in_roots(true);
+}
```

Comment:
I think it might be ok to just have the success test `test_hard_fork_without_gap_in_roots`, without the need to test expecting an explicit failure that we've already fixed. I think this also removes the need for the additional `ValidatorConfig` flag

Reply:
yeah, i made so.